### ISSUE TITLE
Pagination helpers

### ISSUE THREAD

**Issue by martinbonnin** [apollographql/apollo-android#3593]
Using the normalized cache with lists and especially pages is cumbersome as different pages get stored in different records:

```graphql
query GetRepositories($first: Int, $offset: Int) {
  repositories(first: $first, offset: $offset) {
    id
    owner
    name
  }
}
```

This will store the following root Record:

```json
{
  "repositories(first: 100, offset: 0)": [
     { "owner": .. },
     { "owner": .. },
  ], 
  "repositories(first: 100, offset: 100)": [
     { "owner": .. },
     { "owner": .. },
  ], 
}
```

Because the data is spread between two Record fields, there's no easy way for a watcher to retrieve the full list.

API is yet to be defined.

**Comment by sonatard** [comment]
And we need relayStylePagination helper function.
https://www.apollographql.com/docs/react/pagination/cursor-based/#relay-style-cursor-pagination

**Comment by pauldavies83** [comment]
We have a paginated query that returns us a list of our app items, like so…

```
query ItemsQuery($itemArgs: ItemArgs!) {
    items(args: $itemArgs) {
        edges {
            id
            node
        }
        pageInfo
    }
}

type ItemsArgs {
  first: Float
  after: String
  last: Float
  before: String
  filterBy: ItemsFilterByArgs
}
```

where the `pageInfo` contains the pagination info, and `ItemsArgs` allows us to request a “bucket” of data (usually a month of time’s worth).

The use case is for us to regularly sync our items from the server to the normalised cache, and our UI will watch the cache.
It seems the only way for us to do this is to iterate over the pages in the query, and manually `writeOperation` the cache to populate it with all the edges (and key it against the original query with no pagination parameter).

It would be great if there was internal workings in the apolloClient to help us manage this.

**Comment by pauldavies83** [comment]
Hey @martinbonnin 👋 

Did this get prioritised internally? Just wondering if this is something we can expect in the short/medium term?

Thanks! 🙏

**Comment by martinbonnin** [comment]
Yes, it's one of the next big thing on the todo list after https://github.com/apollographql/apollo-kotlin/issues/3566. Pretty hard to commit to a date at this point but the coming months are the current goal.

**Comment by martinbonnin** [comment]
Hi 👋 

We started investigating these APIs. You can read more about it [in the design document](https://github.com/apollographql/apollo-kotlin/blob/main/design-docs/Normalized%20cache%20pagination.md#using-the-incubating-pagination-support). 

The tldr; is:

1. Import the incubating artifact:

```kotlin
// build.gradle.kts

dependencies {
  implementation("com.apollographql.apollo3:apollo-normalized-cache-incubating")
}
```

2. extend your schema:

```graphql
extend type Query @typePolicy(connectionFields: "usersConnection")
```

3. configure your ApolloStore:

```kotlin
val apolloStore = ApolloStore(
  normalizedCacheFactory = cacheFactory,
  cacheKeyGenerator = TypePolicyCacheKeyGenerator,
  metadataGenerator = ConnectionMetadataGenerator(Pagination.connectionTypes),
  apolloResolver = FieldPolicyApolloResolver,
  recordMerger = ConnectionRecordMerger
)
```

Updating the cache now merges new items with the existing ones

4. reading the cache now returns the full non-paginated list.

It's still the early days and too early to use in production but any feedback is warmly welcome.

**Comment by sonatard** [comment]
@martinbonnin Do you have plans to support the following format?

```graphql
type UserConnection {
  pageInfo: PageInfo!
  nodes: [User!]!
}
```

Shopify and GitHub APIs have nodes in addition to edges.
In Modern Relay, Cursors are no longer used, and instead startCursor and endCursor in PageInfo are used. Currently, Edges are an unnecessary layer.

<img width="338" alt="スクリーンショット 2024-03-18 22 08 45" src="https://github.com/apollographql/apollo-kotlin/assets/1586910/07b08a35-6a82-4fd7-a297-a6299875966a">

<img width="401" alt="スクリーンショット 2024-03-18 22 09 14" src="https://github.com/apollographql/apollo-kotlin/assets/1586910/48c8c62d-44aa-4ce3-8b02-82e059e15712">

https://shopify.dev/docs/api/admin-graphql/2024-01/queries/customers#returns

**Comment by BoD** [comment]
Hey @sonatard, this is great feedback! I've created issue #5735 to follow-up on this. Short answer is the current system is flexible enough to manually configure it to work with queries selecting `nodes` instead of `edges` but it would be better if it just worked automatically with the dedicated support we have for Relay style - I think this can be done without too much hurdles.
